mlflow/llama_index/pyfunc_wrapper.py

```diff
@@ -0,0 +1,159 @@
+from typing import Any, Callable, Dict, List, Optional, Union
+
+from llama_index.core import QueryBundle
+from llama_index.core.llms import ChatMessage
+
+from mlflow.models.utils import _convert_llm_input_data
+
+CHAT_ENGINE_NAME = "chat"
+QUERY_ENGINE_NAME = "query"
+RETRIEVER_ENGINE_NAME = "retriever"
+SUPPORTED_ENGINES = {CHAT_ENGINE_NAME, QUERY_ENGINE_NAME, RETRIEVER_ENGINE_NAME}
+
+_CHAT_MESSAGE_PARAMETER_NAMES = {"query", "message"}
+_CHAT_MESSAGE_HISTORY_PARAMETER_NAMES = {"messages", "chat_history"}
+
+
+class _LlamaIndexModelWrapperBase:
+    def __init__(
+        self,
+        index,
+        model_config: Optional[Dict[str, Any]] = None,
+    ):
+        self.index = index
+        self.model_config = model_config or {}
+        self.predict_callable = self._build_engine_method()
+
+    def _build_engine_method(self) -> Callable:
+        raise NotImplementedError
+
+    def _format_predict_input(self, data):
+        raise NotImplementedError
+
+    def _do_inference(self, input, params: Optional[Dict[str, Any]]) -> Dict:
+        """
+        Perform engine inference on a single engine input e.g. not an iterable of
+        engine inputs. The engine inputs must already be preprocessed/cleaned.
+        """
+
+        if isinstance(input, dict):
+            return self.predict_callable(**input, **(params or {}))
+        else:
```

Comment:
Can we add explicit `elif` for the other supported type, and raise for unexpected inputs? (or is it complex - like depending on engine type?)

Reply:
Unfortunately the signature is not consistent, even within a single engine type. I don't see a scalable way to handle unexpected inputs (without essentially recreating python base errors).